src/model_theory/basic.lean

```diff
@@ -366,5 +388,228 @@ lemma comp_assoc (f : M ≃[L] N) (g : N ≃[L] P) (h : P ≃[L] Q) :
 
 end equiv
 
+section sum_Structure
+variables (L₁ L₂ : language) (S : Type*) [L₁.Structure S] [L₂.Structure S]
+
+instance sum_Structure :
+  (L₁.sum L₂).Structure S :=
+{ fun_map := λ n, sum.elim fun_map fun_map,
+  rel_map := λ n, sum.elim rel_map rel_map, }
+
+variables {L₁ L₂ S}
+
+@[simp] lemma fun_map_sum_inl {n : ℕ} (f : L₁.functions n) :
+  @fun_map (L₁.sum L₂) S _ n (sum.inl f) = fun_map f := rfl
+
+@[simp] lemma fun_map_sum_inr {n : ℕ} (f : L₂.functions n) :
+  @fun_map (L₁.sum L₂) S _ n (sum.inr f) = fun_map f := rfl
+
+@[simp] lemma rel_map_sum_inl {n : ℕ} (R : L₁.relations n) :
+  @rel_map (L₁.sum L₂) S _ n (sum.inl R) = rel_map R := rfl
+
+@[simp] lemma rel_map_sum_inr {n : ℕ} (R : L₂.relations n) :
+  @rel_map (L₁.sum L₂) S _ n (sum.inr R) = rel_map R := rfl
+
+end sum_Structure
+
+/-- A language homomorphism maps the symbols of one language to symbols of another. -/
+structure Lhom (L L' : language) :=
+(on_function : ∀{n}, L.functions n → L'.functions n)
+(on_relation : ∀{n}, L.relations n → L'.relations n)
+
+infix ` →ᴸ `:10 := Lhom -- \^L
+
+namespace Lhom
+
+variables (ϕ : L →ᴸ L')
+
+/-- The identity language homomorphism. -/
+protected def id (L : language) : L →ᴸ L :=
+⟨λn, id, λ n, id⟩
+
+instance : inhabited (L →ᴸ L) := ⟨Lhom.id L⟩
+
+/-- The inclusion of the left factor into the sum of two languages. -/
+protected def sum_inl : L →ᴸ L.sum L' :=
+⟨λn, sum.inl, λ n, sum.inl⟩
+
+/-- The inclusion of the right factor into the sum of two languages. -/
+protected def sum_inr : L' →ᴸ L.sum L' :=
+⟨λn, sum.inr, λ n, sum.inr⟩
+
+variables (L L')
+
+/-- The inclusion of an empty language into any other language. -/
+protected def of_is_empty [L.is_algebraic] [L.is_relational] : L →ᴸ L' :=
+⟨λ n, (is_relational.empty_functions n).elim, λ n, (is_algebraic.empty_relations n).elim⟩
+
+variables {L L'}
+
+/-- The composition of two language homomorphisms. -/
+@[reducible] def comp {L1} {L2} {L3} (g : L2 →ᴸ L3) (f : L1 →ᴸ L2) : L1 →ᴸ L3 :=
+begin
+  rcases g with ⟨g1, g2⟩, rcases f with ⟨f1,f2⟩,
+  exact ⟨λn, g1 ∘ f1, λn, g2 ∘ f2⟩,
+end
```

Comment:
```suggestion
⟨λ n, g.1 ∘ f.1, λ n, g.2 ∘ f.2⟩
```
The reason is that in your definition, `(g.comp f).1 x` is not definitionally equal to `g.1 (f.1 x)` if `g` and `f` are variables.

Reply:
That still doesn't work, but I think I have something closer to this.